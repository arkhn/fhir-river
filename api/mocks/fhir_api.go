package mocks

import (
	"fmt"
	"net/http"
	"net/http/httptest"
)

func FhirAPI() *httptest.Server {
	fhirConceptMap := `{
		"id": "cm_gender",
		"group": [{"element": [
			{"code": "F", "target": [{"code": "female", "equivalence": "equal"}]},
			{"code": "M", "target": [{"code": "male", "equivalence": "equal"}]}
		]}]
	}`
	mockFhirApi := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		authorizationHeader := r.Header.Get("Authorization")
		if authorizationHeader == "Bearer validToken" {
			_, _ = fmt.Fprint(w, fhirConceptMap)
		} else if authorizationHeader == "Bearer forbiddenToken" {
			http.Error(w, "invalid token", http.StatusForbidden)
		} else {
			http.Error(w, "invalid token", http.StatusUnauthorized)
		}
	}))

	return mockFhirApi
}
