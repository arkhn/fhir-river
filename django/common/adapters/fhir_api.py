from uuid import uuid4

from django.conf import settings
from django.utils.module_loading import import_string

import requests


class FhirAPI:
    def create(self, resource_type, payload, auth_token=None):
        raise NotImplementedError

    def validate(self, resource_type, payload, auth_token=None):
        raise NotImplementedError

    def retrieve(self, resource_type, resource_id, auth_token=None):
        raise NotImplementedError


class InMemoryFhirAPI(FhirAPI):
    def __init__(self):
        super().__init__()
        self._db = {}

    def create(self, resource_type: str, payload: dict, auth_token=None):
        resource = {"id": uuid4(), **payload}
        if isinstance(self._db.get(resource_type), list):
            self._db[resource_type] += [resource]
        else:
            self._db[resource_type] = [resource]
        return resource

    def validate(self, resource_type: str, payload: dict, auth_token=None):
        return {
            "resourceType": "OperationOutcome",
            "text": {
                "status": "generated",
                "div": "<p> it went fine bro </p>",
            },
            "issue": [],
        }

    def retrieve(self, resource_type, resource_id, auth_token=None):
        for resource in self._db.get(resource_type) or []:
            if resource["id"] == resource_id:
                return resource
        return None


class HapiFhirAPI(FhirAPI):
    def __init__(self):
        super().__init__()
        self._headers = {"Cache-Control": "no-cache", "Content-Type": "application/fhir+json"}
        self._url = settings.FHIR_API_URL

    def create(self, resource_type: str, payload: dict, auth_token=None):
        headers = {**self._headers, "Authorization": f"Bearer {auth_token}"} if auth_token else self._headers
        response = requests.post(
            f"{self._url}/{resource_type}/",
            json=payload,
            headers=headers,
        )
        response.raise_for_status()
        return response.json()

    def validate(self, resource_type: str, payload: dict, auth_token=None):
        """Calls the /<resource_type>/$validate endpoint of HAPI FHIR.
        Note that this function does not raise an exception if the status is not 2XX.

        Args:
            resource_type (str): the resource type
            payload (dict): the FHIR instance
            auth_token ([type], optional): The authentication token to access FHIR API.
            Defaults to None.

        Returns:
            (dict): OperationOutcome containing the details about validation errors.
        """
        headers = {**self._headers, "Authorization": f"Bearer {auth_token}"} if auth_token else self._headers
        response = requests.post(
            f"{self._url}/{resource_type}/$validate",
            json=payload,
            headers=headers,
        )
        try:
            return response.json()
        except Exception:
            response.raise_for_status()

    def retrieve(self, resource_type, resource_id, auth_token=None):
        headers = {**self._headers, "Authorization": f"Bearer {auth_token}"} if auth_token else self._headers

        response = requests.get(
            f"{self._url}/{resource_type}/{resource_id}",
            headers=headers,
        )
        response.raise_for_status()
        return response.json()


fhir_api_class = import_string(settings.DEFAULT_FHIR_API_CLASS)

fhir_api = fhir_api_class()
