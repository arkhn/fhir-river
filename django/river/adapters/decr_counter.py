from typing import Optional

import redis

from django.conf import settings


class DecrementingCounter:
    def set(self, id: str, value: int):
        raise NotImplementedError

    def decr(self, id: str):
        raise NotImplementedError

    def get(self, id: str) -> int:
        raise NotImplementedError


class FakeDecrementingCounter(DecrementingCounter):
    def __init__(self):
        self._count = {}

    def set(self, id: str, value: int):
        self._count = {id: value}

    def decr(self, id: str):
        self._count[id] -= 1

    def get(self, id: str) -> int:
        return self._count[id]


class RedisDecrementingCounter(DecrementingCounter):
    def __init__(self):
        self._client = redis.Redis(
            host=settings.REDIS_HOST,
            port=settings.REDIS_PORT,
            db=settings.REDIS_DB,
        )

    def set(self, id: str, value: int):
        self._client.hset("decr_counters", id, value)

    def decr(self, id: str):
        self._client.hincrby("decr_counters", id, -1)

    def get(self, id: str) -> Optional[int]:
        raw = self._client.hget("decr_counters", id)
        return int(raw) if raw else None
