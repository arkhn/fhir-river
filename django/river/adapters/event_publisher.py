import dataclasses
import json
import logging

import confluent_kafka

from django.conf import settings

from river.domain.events import Event

logger = logging.getLogger(__name__)


class EventPublisher:
    def publish(self, topic: str, event: Event):
        raise NotImplementedError


class FakeEventPublisher(EventPublisher):
    def __init__(self):
        self._events: dict[str, set] = {}

    def publish(self, topic: str, event: Event):
        if topic in self._events:
            self._events[topic].add(event)
        else:
            self._events[topic] = {event}


class KafkaEventPublisher(EventPublisher):
    def __init__(
        self,
    ):
        self._kafka_producer = confluent_kafka.Producer(
            {
                "bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS,
                "linger.ms": 0.5,
                "session.timeout.ms": 6000,
            }
        )

    def publish(self, topic: str, event: Event):
        self._kafka_producer.produce(
            topic=topic,
            value=json.dumps(dataclasses.asdict(event)),
            callback=lambda err, msg: logger.debug("Message delivered")
            if err is None
            else logger.error(f"Message {event} delivery failed with error {err} for topic {msg.topic()}"),
        )
        self._kafka_producer.poll(1)
