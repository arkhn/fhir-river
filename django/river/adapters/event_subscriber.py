import json
import logging
import time
from contextlib import contextmanager
from typing import Any, Dict, List, Tuple

import confluent_kafka
from confluent_kafka import KafkaError, KafkaException

from django.conf import settings

logger = logging.getLogger(__name__)


class EventSubscriber:
    @contextmanager
    def subscribe(self, topic: str):
        raise NotImplementedError

    def poll(self):
        raise NotImplementedError


class FakeEventSubscriber(EventSubscriber):
    def __init__(self, events: Dict[str, List[dict]]):
        self._events = events
        self._seen = []

    @contextmanager
    def subscribe(self, topic: str):
        self.topic = topic
        yield self

    def poll(self):
        curr = self._events[self.topic].pop(0)
        self._seen.append(curr)
        return curr


class KafkaEventSubscriber(EventSubscriber):
    def __init__(self, group_id: str):
        self._kafka_consumer = confluent_kafka.Consumer(
            {
                "bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS,
                "group.id": group_id,
                "linger.ms": 0.5,
                "session.timeout.ms": 6000,
                # topic.metadata.refresh.interval.ms (default 5 min) is the period of
                # time in milliseconds after which we force a refresh of metadata.
                # Here we refresh the list of consumed topics every 5s.
                "topic.metadata.refresh.interval.ms": 5000,
                "auto.offset.reset": "earliest",
            }
        )

    @contextmanager
    def subscribe(self, topics: List[str]):
        self._kafka_consumer.subscribe(topics)
        yield self
        self._kafka_consumer.close()

    def poll(self) -> Tuple[str, Any]:
        """Polls until a message's available for delivery"""
        while True:
            msg = self._kafka_consumer.poll(timeout=1.0)
            if msg is None:
                continue
            if error := msg.error():
                logger.debug(error.str())
                self._handle_error(error)
                continue
            return msg.topic(), json.loads(msg.value())

    def _handle_error(self, error: KafkaError):
        """Handles a KafkaError"""
        if error.retriable() or error.code() == KafkaError.UNKNOWN_TOPIC_OR_PART:
            time.sleep(1)
        else:
            raise KafkaException(error)
