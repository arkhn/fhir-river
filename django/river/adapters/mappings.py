from typing import Any

from common.mapping.fetch_mapping import fetch_resource_mapping
from utils.caching import cache


class MappingUnavailable(Exception):
    pass


def build_repository_key(batch_id: str, resource_id: str):
    return f"{batch_id}:{resource_id}"


class MappingsRepository:
    """This class is an abstraction that is used when we need to fetch a mapping.
    It has a single method `get` with the mapping id as argument.
    """

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        raise NotImplementedError

    def get(self, batch_id: str, resource_id: str):
        raise NotImplementedError


class FakeMappingsRepository(MappingsRepository):
    def __init__(self, mappings: dict = None):
        self._mappings = mappings or {}
        self._seen = []

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        key = build_repository_key(batch_id, resource_id)
        self._mappings[key] = mapping

    def get(self, batch_id: str, resource_id: str):
        key = build_repository_key(batch_id, resource_id)
        self._seen.append(ikeyd)
        return self._mappings[key]


class RedisMappingsRepository(MappingsRepository):
    def __init__(self, mapping_redis: redis.Redis):
        self.mapping_redis = mapping_redis

    def set(self, batch_id: str, resource_id: str, mapping: Any):
        key = build_repository_key(batch_id, resource_id)
        self.mappings_redis.set(key, json.dumps(resource_mapping))

    def get(self, batch_id: str, resource_id: str):
        key = build_repository_key(batch_id, resource_id)
        serialized_mapping = self.mapping_redis.get(key)
        if serialized_mapping is None:
            logger.exception(
                {
                    "message": f"Mapping not found for batch {batch_id} and resource {resource_id}",
                    "resource_id": resource_id,
                },
            )
            raise MappingUnavailable(resource_id)
        mapping = json.loads(serialized_mapping)

        return mapping
