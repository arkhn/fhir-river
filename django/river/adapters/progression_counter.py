from dataclasses import dataclass
from typing import Optional

import redis

from django.conf import settings


@dataclass(frozen=True)
class Progression:
    extracted: int
    loaded: int
    failed: int


class ProgressionCounter:
    """Abstract class used to track the progression of an ETL.

    A `ProgressionCounter` is basically a key-value store.
    The keys should identify a (batch, pyrog source) couple and each key gives access to
    2 values. One of this value is the number of resources extracted for the
    corresponding key and the other is the number of resources currenlty loaded in the
    target DB.

    A `ProgressionCounter` should have 3 methods:
        - set_extracted: sets the value corresponding to the number of extracted
            resources
        - increment_loaded: increments by 1 the value corresponding to the number of
            loaded resources
        - increment_failed: increments by 1 the value corresponding to the number of
            failed resources
        - get: returns both counters for a specified key

    """

    def set_extracted(self, id: str, value: int):
        raise NotImplementedError

    def increment_loaded(self, id: str):
        raise NotImplementedError

    def increment_failed(self, id: str):
        raise NotImplementedError

    def get(self, id: str) -> Optional[Progression]:
        raise NotImplementedError


class FakeProgressionCounter(ProgressionCounter):
    """ProgressionCounter using an in-memory dict as a data structure.

    Attributes:
        _count (dict of str: dict): The dict in which are stored the counters.
        It has the form:
            {
                "key1": {"extracted": nb_extracted_1,
                        "loaded": nb_loaded_1,
                        "failed": nb_failed_2},
                "key2": {"extracted": nb_extracted_2,
                        "loaded": nb_loaded_2,
                        "failed": nb_failed_3},
                ...
            }

    """

    def __init__(self, counts=dict()):
        self._count = counts

    def set_extracted(self, id: str, value: int) -> None:
        """Sets the value corresponding to the number of extracted resources."""
        self._count[id] = {"extracted": value}

    def increment_loaded(self, id: str) -> None:
        """Increments by 1 the value corresponding to the number of loaded resources."""
        if id not in self._count:
            self._count[id] = {}
        if "loaded" not in self._count[id]:
            self._count[id]["loaded"] = 0
        self._count[id]["loaded"] += 1

    def increment_failed(self, id: str) -> None:
        """Increments by 1 the value corresponding to the number of failed resources."""
        if id not in self._count:
            self._count[id] = {}
        if "failed" not in self._count[id]:
            self._count[id]["failed"] = 0
        self._count[id]["failed"] += 1

    def get(self, id: str) -> Optional[Progression]:
        """Gets both counters for the specified key.

        Args:
            id: The key corresponding to the current (batch, resource).

        Returns:
            A tuple with the number of extracted resources (or None if not found) and
                and the number of loaded resources (or None if not found)
                and the number of failed resources (or None if not found).

        """
        counters = self._count.get(id)
        if not counters:
            return None

        extracted = counters.get("extracted")
        loaded = counters.get("loaded")
        failed = counters.get("failed")

        return Progression(extracted=extracted, loaded=loaded, failed=failed)


class RedisProgressionCounter(ProgressionCounter):
    """ProgressionCounter using two Redis hashes to record extract and load

    Attributes:
        _client (redis.Redis): The redis client used to access the backend redis DB.

    """

    def __init__(self):
        self._client = redis.Redis(host=settings.REDIS_HOST, port=settings.REDIS_PORT, db=settings.REDIS_DB)

    def set_extracted(self, id: str, value: int) -> None:
        """Sets the value corresponding to the number of extracted resources."""
        self._client.hset("extracted_counters", id, value)

    def increment_loaded(self, id: str) -> None:
        """Increments by 1 the value corresponding to the number of loaded resources."""
        self._client.hincrby("loaded_counters", id, 1)

    def increment_failed(self, id: str) -> None:
        """Increments by 1 the value corresponding to the number of failed resources."""
        self._client.hincrby("failed_counters", id, 1)

    def get(self, id: str) -> Optional[Progression]:
        """Gets both counters for the specified key.

        Args:
            id: The key corresponding to the current (batch, resource).

        Returns:
            A tuple with the number of extracted resources (or None if not found)
                and the number of loaded resources (or None if not found)
                and the number of failed resources (or None if not found).

        """
        raw_extracted = self._client.hget("extracted_counters", id)
        raw_loaded = self._client.hget("loaded_counters", id)
        raw_failed = self._client.hget("failed_counters", id)
        extracted = int(raw_extracted) if raw_extracted else None
        loaded = int(raw_loaded) if raw_loaded else None
        failed = int(raw_failed) if raw_failed else None

        return Progression(extracted=extracted, loaded=loaded, failed=failed)
