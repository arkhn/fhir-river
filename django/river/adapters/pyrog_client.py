from river.common.mapping.fetch_mapping import fetch_resource_mapping

# FIXME this shouldn't be needed when the DB is shared


class PyrogClient:
    """This class is an abstraction that is used when we need to fetch a mapping.
    It has a single method `get` with the mapping id as argument.
    """

    def fetch_mapping(self, resource_id: str):
        raise NotImplementedError


class FakePyrogClient(PyrogClient):
    def __init__(self, mappings: dict = None):
        self._mappings = mappings or {}
        self._seen = []

    def fetch_mapping(self, resource_id: str):
        if resource_id in self._mappings:
            self._seen.append(resource_id)
        return self._mappings.get(resource_id)


class APIPyrogClient(PyrogClient):
    def __init__(self, authorization_header: str):
        self.authorization_header = authorization_header

    def fetch_mapping(self, resource_id: str):
        return fetch_resource_mapping(resource_id, self.authorization_header)
