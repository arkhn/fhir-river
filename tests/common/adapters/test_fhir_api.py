from common.adapters.fhir_api import HapiFhirAPI


def test_fhir_api():
    fhir_api = HapiFhirAPI()

    assert fhir_api._headers == {"Cache-Control": "no-cache", "Content-Type": "application/fhir+json"}
