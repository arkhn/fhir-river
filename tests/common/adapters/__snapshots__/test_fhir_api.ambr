# name: test_hapi_fhir_api_validate_error
  <class 'dict'> {
    'issue': <class 'list'> [
      <class 'dict'> {
        'code': '',
        'diagnostics': '500 internal server error',
        'severity': 'error',
      },
    ],
    'resourceType': 'OperationOutcome',
  }
---
# name: test_inmemory_fhir_api_validate
  <class 'dict'> {
    'issue': <class 'list'> [
    ],
    'resourceType': 'OperationOutcome',
    'text': <class 'dict'> {
      'div': '<p> it went fine bro </p>',
      'status': 'generated',
    },
  }
---
